File: src/main/scala/com/singhraghav/ds/stack/ArrayStack.scala
package com.singhraghav.ds.stack

import scala.reflect.ClassTag

class ArrayStack[T] (private val data: Array[T], private var capacity: Int, private var t: Int = -1) extends MyStack[T] {
  override def push(element: T): ArrayStack.this.type = {
    if (size() == data.length) throw new IllegalStateException("Stack full")
    t = t + 1
    data(t) = element
    this
  }

  override def pop(): T = {
    if (isEmpty()) throw new NoSuchElementException("Stack Empty")
    else {
      val elementToReturn = data(t)
      t -= 1
      elementToReturn
    }
  }

  override def isEmpty(): Boolean = t == -1

  override def size(): Int = t + 1

  override def top(): T =
    if (isEmpty()) throw new NoSuchElementException("Stack Empty") else data(t)
}

object ArrayStack {

  def apply[T: ClassTag](): MyStack[T] = apply(1000)

  def apply[T: ClassTag](capacity: Int)= new ArrayStack[T](data = Array.ofDim[T](capacity), capacity = capacity)
}

File: src/main/java/com/singraghav/ds/stack/ArrayStack.java
package com.singraghav.ds.stack;

public class ArrayStack<E> implements MStack<E> {

    public static final int CAPACITY = 100;

    private E[] data;

    private int t = -1;

    public ArrayStack() {
        this(CAPACITY);
    }

    public ArrayStack(int capacity) {
        data = (E[]) new Object [capacity];
    }

    @Override
    public void push(E element) {
        if (size() == data.length) throw new IllegalStateException("Stack full");
        data[++t] = element;
    }

    @Override
    public E pop() {
        if (isEmpty()) return null;
        E top = data[t];
        data[t] = null;
        t -= 1;
        return top;
    }

    @Override
    public E top() {
        if (isEmpty()) return null;
        return data[t];
    }

    @Override
    public int size() {
        return t + 1;
    }

    @Override
    public boolean isEmpty() {
        return t == -1;
    }
}
